input/parser.go

```diff
@@ -2,68 +2,80 @@ package input
 
 import (
 	"fmt"
-	"strings"
+	"github.com/ekanite/ekanite/parser"
+	"log"
 )
 
-var (
-	fmtsByStandard = []string{"rfc5424"}
-	fmtsByName     = []string{"syslog"}
+const (
+	Rfc5424Standard   = "RFC5424"
+	Rfc5424Name       = "syslog"
+	WatchguardFirebox = "Watchguard"
+	WatchguardName    = "M200"
 )
 
+type StatsCollector func(key string, delta int64)
+
+type LogParser interface {
+	Parse(raw []byte, result *map[string]interface{})
+	CompileMatcher()
+}
+
+// A Parser parses the raw input as a map with a timestamp field.
+type LogHandler struct {
+	Fmt    string
+	Raw    []byte
+	Result map[string]interface{}
+	Parser LogParser
+	Stats func(key string, delta int64)
+}
+
+func supportedFormats() [][]string {
+	return [][]string{{Rfc5424Name, Rfc5424Standard},
+		{WatchguardName, WatchguardFirebox}}
+}
+
 // ValidFormat returns if the given format matches one of the possible formats.
-func ValidFormat(format string) bool {
-	for _, f := range append(fmtsByStandard, fmtsByName...) {
-		if f == format {
+func ValidFormat(f string) bool {
+
+	l := len(supportedFormats())
+	fmts := supportedFormats()
+
+	for i := 0; i < l; i++ {
+		if fmts[i][0] == f {
 			return true
 		}
 	}
-	return false
-}
 
-// A Parser parses the raw input as a map with a timestamp field.
-type Parser struct {
-	fmt     string
-	Raw     []byte
-	Result  map[string]interface{}
-	rfc5424 *RFC5424
+	return false
 }
 
 // NewParser returns a new Parser instance.
-func NewParser(f string) (*Parser, error) {
+func NewParser(f string) (*LogHandler, error) {
 	if !ValidFormat(f) {
-		return nil, fmt.Errorf("%s is not a valid format", f)
+		return nil, fmt.Errorf("%s is not a valid parser format", f)
 	}
 
-	p := &Parser{}
-	p.detectFmt(strings.TrimSpace(strings.ToLower(f)))
-	p.newRFC5424Parser()
-	return p, nil
-}
+	var p = &LogHandler{}
 
-// Reads the given format and detects its internal name.
-func (p *Parser) detectFmt(f string) {
-	for i, v := range fmtsByName {
-		if f == v {
-			p.fmt = fmtsByStandard[i]
-			return
-		}
-	}
-	for _, v := range fmtsByStandard {
-		if f == v {
-			p.fmt = v
-			return
-		}
+	if f == Rfc5424Name {
+		p.Parser = &parser.RFC5424{}
+	} else if f == WatchguardName {
+		p.Parser = &parser.Watchguard{}
+	} else {
+		panic(fmt.Sprintf("no supported parser for input format %s", f))
```

Comment:
This should return an error, not panic. However it's not even sure you can here since the code checks validity earlier.